File: Day80/Part4/combo_decay.py
import sys
from collections import Counter

def solve():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    s = data[0]
    counts = Counter(s)
    
    perm = []
    
    while counts:
        # Get one of each available character, sorted
        layer = sorted(counts.keys())
        perm.extend(layer)
        
        # Decrement counts
        to_remove = []
        for char in layer:
            counts[char] -= 1
            if counts[char] == 0:
                to_remove.append(char)
        
        for char in to_remove:
            del counts[char]
            
    total_score = 0
    current_combo = 0
    
    if perm:
        current_combo = 1
        total_score += current_combo
        
        for i in range(1, len(perm)):
            if perm[i] > perm[i-1]:
                current_combo += 1
            else:
                current_combo //= 2
            total_score += current_combo
            
    print(total_score)

if __name__ == "__main__":
    solve()

File: Day80/Part4/combo_decay.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

int main() {
    string s;
    if (!(cin >> s)) return 0;

    map<char, int> counts;
    for (char c : s) {
        counts[c]++;
    }

    string perm = "";
    while (!counts.empty()) {
        string layer = "";
        vector<char> to_remove;
        for (auto& pair : counts) {
            layer += pair.first;
            pair.second--;
            if (pair.second == 0) {
                to_remove.push_back(pair.first);
            }
        }
        for (char c : to_remove) {
            counts.erase(c);
        }
        // Layer is already sorted because map iterates in order
        perm += layer;
    }

    long long total_score = 0;
    long long current_combo = 0;

    if (!perm.empty()) {
        // First char
        current_combo = 1;
        total_score += current_combo;

        for (size_t i = 1; i < perm.length(); ++i) {
            if (perm[i] > perm[i-1]) {
                current_combo++;
            } else {
                current_combo /= 2;
            }
            total_score += current_combo;
        }
    }

    cout << total_score << endl;

    return 0;
}
